[Skia] Add support for favicons on GTK

TingPing opened the pull request:
#### e9c4929afd39ec66bf6c99028c2445be9bfd1522
<pre>
[Skia] Add support for favicons on GTK
<a href="https://bugs.webkit.org/show_bug.cgi?id=270811">https://bugs.webkit.org/show_bug.cgi?id=270811</a>

Reviewed by Carlos Garcia Campos.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp: Added.
(WebCore::skiaImageToGdkTexture):
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_favicon_finish):

Canonical link: <a href="https://commits.webkit.org/275982@main">https://commits.webkit.org/275982@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc9cc536d547dde1777707d358c30b2ee87e8c50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46073 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45742 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19887 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35911 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19508 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37417 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16887 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38480 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1494 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38789 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47617 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18444 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15102 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42694 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19890 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41359 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20070 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5907 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+#if USE(GTK4)
+GRefPtr<GdkTexture> skiaImageToGdkTexture(SkImage* image)
+{
+    ASSERT(image);
```
Pass a reference and we don't need the assert here.

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+
+cairo_surface_t* skiaImageToCairoSurface(SkImage* image)
+{
+    ASSERT(image);
```
Ditto.

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+        return nullptr;
+
+    auto* data = static_cast<const uint8_t*>(pixmap.addr());
+    auto* surface = cairo_image_surface_create_for_data(const_cast<uint8_t*>(data), CAIRO_FORMAT_ARGB32, pixmap.width(), pixmap.height(), pixmap.rowBytes());
```
You can use `pixmap.writable_addr8()`

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+    auto* data = static_cast<const uint8_t*>(pixmap.addr());
+    auto* surface = cairo_image_surface_create_for_data(const_cast<uint8_t*>(data), CAIRO_FORMAT_ARGB32, pixmap.width(), pixmap.height(), pixmap.rowBytes());
+    if (cairo_surface_status(surface) != CAIRO_STATUS_SUCCESS) {
+        cairo_surface_destroy(surface);
```
You can use RefPtr, for the GTK port it's build with Skia enabled too.

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+    if (status != CAIRO_STATUS_SUCCESS) {
+        cairo_surface_destroy(surface);
+        return nullptr;
+    }
```
We don't check failure of set_user_data in other places, I think this can only fail if an invalid surface is passed or in case of OOM

TingPing replied:
Yeah it’s for OOM. Might as well `g_error()` then?

carlosgcampos replied:
I think we can just ignore it like we do everywhere else

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,86 @@
+
+#else
+
+cairo_surface_t* skiaImageToCairoSurface(SkImage* image)
```
This should return `RefPtr<cairo_surface_t>`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -254,18 +256,25 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+#else
+#if USE(SKIA)
+    auto* image = static_cast<SkImage*>(g_task_propagate_pointer(G_TASK(result), error));
+    return image ? skiaImageToCairoSurface(image) : nullptr;
```
I see here it's more convenient to return the raw pointer, but we might need this helper in other places like snapshots, so better be consistent with `skiaImageToGdkTexture`

carlosgcampos reviewed (approved):
Excellent, thanks! I have only a few minor comments

carlosgcampos commented on Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp:
```diff
@@ -0,0 +1,74 @@
+
+#else
+
+GRefPtr<cairo_surface_t> skiaImageToCairoSurface(SkImage& image)
```
cairo_surface_t is not a GObject, this should be RefPtr and RefPtrCairo.h needs to be included

TingPing committed:
[Skia] Add support for favicons on GTK
https://bugs.webkit.org/show_bug.cgi?id=270811

Reviewed by Carlos Garcia Campos.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.cpp: Added.
(WebCore::skiaImageToGdkTexture):
* Source/WebCore/platform/graphics/gtk/GdkSkiaUtilities.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_favicon_finish):

Canonical link: https://commits.webkit.org/275982@main
